src/Factory/AbstractBusFactory.php

```diff
@@ -0,0 +1,142 @@
+<?php
+/*
+ * This file is part of the prooph/service-bus.
+ * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ * 
+ * Date: 30.10.14 - 12:57
+ */
+
+namespace Prooph\ServiceBus\Factory;
+
+use Interop\Container\ContainerInterface;
+use Prooph\Common\Messaging\MessageFactory;
+use Prooph\ServiceBus\Exception\RuntimeException;
+use Prooph\ServiceBus\MessageBus;
+use Prooph\ServiceBus\Plugin\MessageFactoryPlugin;
+use Prooph\ServiceBus\Plugin\ServiceLocatorPlugin;
+
+/**
+ * Class AbstractBusFactory
+ *
+ * @package Prooph\ServiceBus\Factory
+ * @author Alexander Miertsch <kontakt@codeliner.ws>
+ */
+abstract class AbstractBusFactory
+{
+    /**
+     * Returns the FQCN of a bus extending Prooph\ServiceBus\MessageBus
+     *
+     * @return string
+     */
+    abstract protected function getBusClass();
+
+    /**
+     * Returns config key used within the prooph.service_bus config namespace to identify the bus.
+     *
+     * @return string
+     */
+    abstract protected function getBusConfigKey();
+
+    /**
+     * Returns the default router class to use if no one was specified in the config
+     *
+     * @return string
+     */
+    abstract protected function getDefaultRouterClass();
+
+    /**
+     * Create service
+     *
+     * @param ContainerInterface $container
+     * @throws RuntimeException
+     * @return MessageBus
+     */
+    public function __invoke(ContainerInterface $container)
+    {
+        $busConfig = $this->getBusConfig($container);
+
+        $busClass = $this->getBusClass();
+
+        $bus = new $busClass();
+
+        if (isset($busConfig['plugins'])) {
+            $this->attachPlugins($bus, $busConfig['plugins'], $container);
+        }
+
+        if (isset($busConfig['router'])) {
+            $this->attachRouter($bus, $busConfig['router']);
+        }
+
+        if (!isset($busConfig['enable_handler_location']) || (bool)$busConfig['enable_handler_location']) {
+            $bus->utilize(new ServiceLocatorPlugin($container));
+        }
+
+        $messageFactoryServiceId = isset($busConfig['message_factory'])
+            ? (string)$busConfig['message_factory'] : MessageFactory::class;
+
+        if ($container->has($messageFactoryServiceId)) {
+            $bus->utilize(new MessageFactoryPlugin($container->get($messageFactoryServiceId)));
+        }
+
+        return $bus;
+    }
+
+    /**
+     * @param ContainerInterface $container
+     * @return array
+     * @throws RuntimeException
+     */
+    private function getBusConfig(ContainerInterface $container)
+    {
+        if (! $container->has('config')) {
+            return [];
+        }
+
+        $config = (array)$container->get('config');
```

Comment:
Is this really neccessary? If container is an object, it should implement array access.